src/undate/date.py

```diff
@@ -0,0 +1,140 @@
+    # adapted from https://stackoverflow.com/a/27129510/9706217
+
+    def __new__(
+        cls,
+        year: Union[int, np.datetime64],
+        month: Optional[int] = None,
+        day: Optional[int] = None,
+    ):
+        if isinstance(year, np.datetime64):
+            _data = year
+        else:
+            datestr = str(year)
+            if month is not None:
+                datestr = f"{year}-{month:02d}"
+                if day is not None:
+                    datestr = f"{datestr}-{day:02d}"
+            _data = np.datetime64(datestr)
+
+        data = np.asarray(_data, dtype="datetime64")
+
+        # expected dtype depends on date unit / how much of date is known
+        expected_unit = "Y"
+        if day is not None and month is not None:
+            expected_unit = "D"
+        elif month:
+            expected_unit = "M"
+        expected_dtype = f"datetime64[{expected_unit}]"
+
+        if data.dtype != expected_dtype:
+            raise ValueError(
+                f"Unable to parse dates adequately as {expected_dtype}: {data}"
+            )
+        obj = data.view(cls)
+        return obj
+
+    def Export(self):
+        return self
+
+    def __array_finalize__(self, obj):
+        if obj is None:
+            return
+
+    # custom properties to access year, month, day
+
+    @property
+    def year(self):
+        return int(str(self.astype("datetime64[Y]")))
+
+    @property
+    def month(self):
+        # if date unit is year, don't return a month (only M/D)
+        if self.dtype != "datetime64[Y]":
+            return int(str(self.astype("datetime64[M]")).split("-")[-1])
+
+    @property
+    def day(self):
+        # only return a day if date unit is in days
+        if self.dtype == "datetime64[D]":
+            return int(str(self.astype("datetime64[D]")).split("-")[-1])
+
+    def __sub__(self, other):
+        # modify to conditionally return a timedelta object instead of a
+        # Date object with dtype timedelta64[D] (default behavior)
+
+        result = super().__sub__(other)
+        # if the result has a timedelta type (i.e., date minus date = timedelta),
+        # cast to local Timedelta object; otherwise, leave as is
+        # (i.e., date minus timedelta = date)
+        if result.dtype == "timedelta64[D]":
+            result = Timedelta(result)
+        return result
+
+    # NOTE: add should not be subclassed because we want to return a Date, not a delta
+
+
+class DatePrecision(IntEnum):
+    """date precision, to indicate date precision independent from how much
+    of the date is known."""
+
+    # NOTE: values MUST be ordered based on the relative size or
+    # precison of the time unit. That is, the smaller the unit, the more precise
+    # it is: a day is more precise than a month, a month is more precise than a year,
+    # (DatePrecision.year < DatePrecision.month)
+
+    #: year
+    YEAR = 1
+    #: month
+    MONTH = 2
+    #: day
+    DAY = 3
```

Comment:
One potential issue with ordering it this way is that it's going to be harder to extend. I guess we could do `DECADE = 0`? My understanding is that we can't get precise than `DAY` (since we're not recording times) but there are potentially future cases where we might want to be less precise than `YEAR`?

Reply:
I was thinking the numbers would change as long as they are sequenced properly. The numbers only need to be consistent within a given installation of undate, right? (we're not storing them in a db or anything).

We could add decade and century now if you think it would be helpful, but was waiting until we had use cases.

Reply:
I was mostly thinking of the conversation earlier today where we heard that even decade precision is still very useful and even century is much better than nothing.

If we were to update this kind of core config it should probably be in at least a minor release? Also - is this "the PR" for v1.0, or at least v0.3? Looks like the last release was April?

Reply:
I'll add century and decade to try to future proof a little bit, but I don't think the actual numbers should matter - they should never be referenced as numbers, only as constants, and they are only defined within the context of an installed version of undate. Am I thinking about that wrong?